File: p349e/intersection.py
from typing import List


class Solution:
    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        # return list(set(nums1) & set(nums2))
        set1 = set(nums1)
        set2 = set([n for n in nums2 if n in set1])
        return list(set2)


# TESTS
tests = [
    [[], []],
    [[1], [2]],
    [[1, 2, 2], [3]],
    [[1, 2, 2, 1], [2, 2]],
    [[4, 9, 5], [9, 4, 9, 8, 4]],
]
for t in tests:
    sol = Solution()
    print(
        "Intersection of", t[0], "and", t[1], "->", sol.intersection(t[0], t[1])
    )

File: p349e/intersection.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var intersection = function(nums1, nums2) {
  let set1 = new Set(nums1);
  return [...new Set(nums2.filter(n => set1.has(n)))];
};
// TESTS
[
  [[], []],
  [[1], [2]],
  [[1, 2, 2], [3]],
  [
    [1, 2, 2, 1],
    [2, 2]
  ],
  [
    [4, 9, 5],
    [9, 4, 9, 8, 4]
  ]
].forEach(t => {
  console.log(
    'Intersection of',
    t[0],
    'and',
    t[1],
    '->',
    intersection(t[0], t[1])
  );
});
